File: 344-reverse-string/344-reverse-string.cpp

//MY Method Time=O(N), Space = O(N)(recursion stack)
class Solution {
public:
    void reverseString(vector<char>& s) {
       if(s.empty())
           return;
        char ele = s.back();
        s.pop_back();
        reverseString(s);
        s.insert(s.begin(),ele);
    }
};

//Using two pointer => Time= O(N)(total N/2 swaps), Space = O(N)(recursion stack)
// class Solution {
// public:
//     void reverseString(vector<char>& s) {
//        helper(s,0,s.size()-1);
//     }
//     void helper(vector<char>& s,int left,int right)
//     {
//         if(left>=right)
//             return;
//         char temp = s[left];
//         s[left++] = s[right];
//         s[right--] = temp;
//         helper(s,left,right);
//     }
// };

//BEST APPROACH
//Using Two Pointers but not Recursion => Time = O(N), Space = O(1)(since no extra space used done inplace)
// class Solution {
// public:
//     void reverseString(vector<char>& s) {
//         int left = 0;
//         int right = s.size()-1;
//         while(left<right)
//         {
//             char temp = s[left];
//             s[left++] = s[right];
//             s[right--] = temp;
//         }
//     }
// };

File: 344-reverse-string/344-reverse-string.java
//RECURSIVE SOLUTION TO REVERSE LIST USING TWO POINTERS AND SWAPPING THEM RECURSIVELY
class Solution {
    public void reverseString(char[] s) {
        reverse(s,0,s.length-1);
    }
    private void reverse(char[] s,int left, int right)
    {
        if(left>=right)
            return;
        char temp = s[left];
        s[left] = s[right];
        s[right] = temp;
        reverse(s,++left,--right);        
    }
}

//BUT HERE NORMAL SWAPPING USING TWO POINTERS WITHOUT USING RECURSION IS BETTER SOLUTION SINCE RECURSION USES STACK